File: 12. Smallest Substring Containing All Characters.py
"""
Smallest Substring Containing All Characters of Another String
"""
from collections import Counter

def min_window(s, t):
    need = Counter(t)
    missing = len(t)
    left = start = 0
    minlen = float('inf')
    for right, ch in enumerate(s):
        if need[ch] > 0: missing -= 1
        need[ch] -= 1
        while missing == 0:
            if right-left+1 < minlen:
                minlen = right-left+1; start = left
            need[s[left]] += 1
            if need[s[left]] > 0: missing += 1
            left += 1
    return '' if minlen==float('inf') else s[start:start+minlen]

def main():
    print(min_window('ADOBECODEBANC','ABC'))

if __name__ == '__main__':
    main()

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 12. Smallest Substring Containing All Characters.cpp
// Topic: Smallest Substring Containing All Characters of Another String
#include <bits/stdc++.h>
using namespace std;

int main()
{
    string s = "ADOBECODEBANC", t = "ABC";
    vector<int> need(128, 0);
    for (char c : t) need[c]++;
    int missing = t.size(), left = 0, start = 0, minLen = INT_MAX;
    for (int right = 0; right < s.size(); right++)
    {
        if (need[s[right]]-- > 0)
            missing--;
        while (missing == 0)
        {
            if (right - left + 1 < minLen)
            {
                minLen = right - left + 1;
                start = left;
            }
            if (++need[s[left++]] > 0)
                missing++;
        }
    }
    cout << (minLen == INT_MAX ? string("") : s.substr(start, minLen)) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
